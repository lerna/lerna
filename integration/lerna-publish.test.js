"use strict";

const execa = require("execa");
const fs = require("fs-extra");
const globby = require("globby");
const normalizeNewline = require("normalize-newline");
const writeJsonFile = require("write-json-file");
const loadJsonFile = require("load-json-file");
const path = require("path");
const os = require("os");
const tempy = require("tempy");

const cliRunner = require("@lerna-test/cli-runner");
const gitAdd = require("@lerna-test/git-add");
const gitCommit = require("@lerna-test/git-commit");
const gitTag = require("@lerna-test/git-tag");
const cloneFixture = require("@lerna-test/clone-fixture")(
  path.resolve(__dirname, "../commands/publish/__tests__")
);
const loadManifests = require("@lerna-test/load-manifests");
const normalizeTestRoot = require("@lerna-test/normalize-test-root");

// stabilize changelog commit SHA and datestamp
expect.addSnapshotSerializer(require("@lerna-test/serialize-changelog"));

const lastCommitMessage = cwd =>
  execa.stdout("git", ["log", "-1", "--format=%B"], { cwd }).then(normalizeNewline);

async function commitChangeToPackage(cwd, packageName, commitMsg, data) {
  const packageJSONPath = path.join(cwd, "packages", packageName, "package.json");
  const pkg = await loadJsonFile(packageJSONPath);

  await writeJsonFile(packageJSONPath, Object.assign(pkg, data));
  await gitAdd(cwd, packageJSONPath);

  return gitCommit(cwd, commitMsg);
}

describe("lerna publish", () => {
  const currentDirectory = process.cwd();

  afterEach(() => {
    // conventional-recommended-bump is incapable of accepting cwd config :P
    if (process.cwd() !== currentDirectory) {
      process.chdir(currentDirectory);
    }
  });

  test("exit 0 when no updates", async () => {
    const { cwd } = await cloneFixture("normal");
    const args = ["publish"];

    await gitTag(cwd, "v1.0.0");

    const { code, stdout } = await cliRunner(cwd)(...args);

    expect(code).toBe(0);
    expect(stdout).toBe("");
  });

  test("updates fixed versions", async () => {
    const { cwd } = await cloneFixture("normal");
    const args = ["publish", "--skip-npm", "--cd-version=patch", "--yes"];

    const { stdout } = await cliRunner(cwd)(...args);
    expect(stdout).toMatchSnapshot("stdout");

    const [allPackageJsons, commitMessage] = await Promise.all([loadManifests(cwd), lastCommitMessage(cwd)]);

    expect(allPackageJsons).toMatchSnapshot("packages");
    expect(commitMessage).toMatchSnapshot("commit");
  });

  test("updates all transitive dependents", async () => {
    const { cwd } = await cloneFixture("snake-graph");
    const args = ["publish", "--skip-npm", "--cd-version=major", "--yes"];

    await gitTag(cwd, "v1.0.0");
    await commitChangeToPackage(cwd, "package-1", "change", { change: true });

    await cliRunner(cwd)(...args);

    expect(await loadManifests(cwd)).toMatchSnapshot();
  });

  test("uses default suffix with canary flag", async () => {
    const { cwd } = await cloneFixture("normal");
    const args = ["publish", "--canary", "--skip-npm", "--yes"];

    const { stdout } = await cliRunner(cwd)(...args);
    expect(stdout).toMatchSnapshot("stdout");
  });

  test("updates independent versions", async () => {
    const { cwd } = await cloneFixture("independent");
    const args = ["publish", "--skip-npm", "--cd-version=major", "--yes"];

    const { stdout } = await cliRunner(cwd)(...args);
    expect(stdout).toMatchSnapshot("stdout");

    const [allPackageJsons, commitMessage] = await Promise.all([loadManifests(cwd), lastCommitMessage(cwd)]);

    expect(allPackageJsons).toMatchSnapshot("packages");
    expect(commitMessage).toMatchSnapshot("commit");
  });

  ["normal", "independent"].forEach(flavor =>
    test(`${flavor} mode --conventional-commits changelog`, async () => {
      const { cwd } = await cloneFixture(`${flavor}`, "feat: init repo");
      const args = ["publish", "--conventional-commits", "--skip-git", "--skip-npm", "--yes"];

      await commitChangeToPackage(cwd, "package-1", "feat(package-1): Add foo", { foo: true });
      await commitChangeToPackage(cwd, "package-1", "fix(package-1): Fix foo", { foo: false });
      await commitChangeToPackage(cwd, "package-2", "fix(package-2): Fix bar", { bar: true });
      await commitChangeToPackage(
        cwd,
        "package-3",
        `feat(package-3): Add baz feature${os.EOL}${os.EOL}BREAKING CHANGE: yup`,
        { baz: true }
      );

      // conventional-recommended-bump is incapable of accepting cwd config :P
      process.chdir(cwd);

      const { stdout } = await cliRunner(cwd)(...args);
      expect(stdout).toMatchSnapshot();

      const changelogFilePaths = await globby(["**/CHANGELOG.md"], {
        cwd,
        absolute: true,
        followSymlinkedDirectories: false,
      });
      const changelogContents = await Promise.all(
        changelogFilePaths.sort().map(fp => fs.readFile(fp, "utf8"))
      );

      expect(changelogContents).toMatchSnapshot();
    })
  );

  it("replaces file: specifier with local version before npm publish but after git commit", async () => {
    const { cwd } = await cloneFixture("relative-file-specs");

    await gitTag(cwd, "v1.0.0");
    await commitChangeToPackage(cwd, "package-1", "feat(package-1): changed", { changed: true });

    await cliRunner(cwd)("publish", "--cd-version=major", "--skip-npm", "--yes");

    expect(
      await execa.stdout("git", ["show", "--unified=0", "--ignore-space-at-eol", "--format=%s"], { cwd })
    ).toMatchSnapshot("committed");

    expect(
      await execa.stdout("git", ["diff", "--unified=0", "--ignore-space-at-eol"], { cwd })
    ).toMatchSnapshot("unstaged");
  });

  test("calls lifecycle scripts", async () => {
    const { cwd } = await cloneFixture("lifecycle");
    const args = ["publish", "--skip-npm", "--cd-version", "minor", "--yes"];

    const { stdout } = await cliRunner(cwd)(...args);
    expect(normalizeTestRoot(cwd)(stdout)).toMatchSnapshot();
  });

  test("silences lifecycle scripts with --loglevel=silent", async () => {
    const { cwd } = await cloneFixture("lifecycle");
    const args = ["publish", "--skip-npm", "--cd-version", "minor", "--yes", "--loglevel", "silent"];

    const { stdout } = await cliRunner(cwd)(...args);
    expect(normalizeTestRoot(cwd)(stdout)).toMatchSnapshot();
  });

  test("exits with EBEHIND when behind upstream remote", async () => {
    const { cwd, repository } = await cloneFixture("normal");
    const cloneDir = tempy.directory();

    // simulate upstream change from another clone
    await execa("git", ["clone", repository, cloneDir]);
    await fs.outputFile(path.join(cloneDir, "README.md"), "upstream change");
    await gitAdd(cloneDir, "-A");
    await gitCommit(cloneDir, "upstream change");
    await execa("git", ["push", "origin", "master"], { cwd: cloneDir });

    await expect(cliRunner(cwd)("publish")).rejects.toThrowError(/EBEHIND/);
  });
});
