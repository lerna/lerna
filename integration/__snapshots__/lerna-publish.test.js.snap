// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lerna publish calls lifecycle scripts 1`] = `

Changes:
 - package-1: 1.0.0 => 1.1.0
 - package-2: 1.0.0 => 1.1.0


> lifecycle@0.0.0-monorepo preversion __TEST_ROOTDIR__
> echo preversion-root

preversion-root

> package-1@1.0.0 preversion __TEST_ROOTDIR__/packages/package-1
> echo preversion-package-1

preversion-package-1

> package-1@1.1.0 version __TEST_ROOTDIR__/packages/package-1
> echo version-package-1

version-package-1

> lifecycle@0.0.0-monorepo version __TEST_ROOTDIR__
> echo version-root

version-root

> package-1@1.1.0 postversion __TEST_ROOTDIR__/packages/package-1
> echo postversion-package-1

postversion-package-1

> lifecycle@0.0.0-monorepo postversion __TEST_ROOTDIR__
> echo postversion-root

postversion-root
`;

exports[`lerna publish independent mode --conventional-commits changelog 1`] = `

Changes:
 - package-1: 1.0.0 => 1.1.0
 - package-2: 2.0.0 => 2.1.0
 - package-3: 3.0.0 => 4.0.0
 - package-4: 4.0.0 => 4.1.0
 - package-5: 5.0.0 => 5.1.0 (private)

`;

exports[`lerna publish independent mode --conventional-commits changelog 2`] = `
Array [
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.1.0"></a>
# 1.1.0 (YYYY-MM-DD)


### Bug Fixes

* **package-1:** Fix foo SHA


### Features

* init repo SHA
* **package-1:** Add foo SHA
,
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="2.1.0"></a>
# 2.1.0 (YYYY-MM-DD)


### Bug Fixes

* **package-2:** Fix bar SHA


### Features

* init repo SHA
,
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="4.0.0"></a>
# 4.0.0 (YYYY-MM-DD)


### Features

* init repo SHA
* **package-3:** Add baz feature SHA


### BREAKING CHANGES

* **package-3:** yup
,
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="4.1.0"></a>
# 4.1.0 (YYYY-MM-DD)


### Features

* init repo SHA
,
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="5.1.0"></a>
# 5.1.0 (YYYY-MM-DD)


### Features

* init repo SHA
,
]
`;

exports[`lerna publish normal mode --conventional-commits changelog 1`] = `

Changes:
 - package-1: 1.0.0 => 2.0.0
 - package-2: 1.0.0 => 2.0.0
 - package-3: 1.0.0 => 2.0.0
 - package-4: 1.0.0 => 2.0.0
 - package-5: 1.0.0 => 2.0.0 (private)

`;

exports[`lerna publish normal mode --conventional-commits changelog 2`] = `
Array [
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="2.0.0"></a>
# 2.0.0 (YYYY-MM-DD)


### Bug Fixes

* **package-1:** Fix foo SHA
* **package-2:** Fix bar SHA


### Features

* init repo SHA
* **package-1:** Add foo SHA
* **package-3:** Add baz feature SHA


### BREAKING CHANGES

* **package-3:** yup
,
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="2.0.0"></a>
# 2.0.0 (YYYY-MM-DD)


### Bug Fixes

* **package-1:** Fix foo SHA


### Features

* init repo SHA
* **package-1:** Add foo SHA
,
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="2.0.0"></a>
# 2.0.0 (YYYY-MM-DD)


### Bug Fixes

* **package-2:** Fix bar SHA


### Features

* init repo SHA
,
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="2.0.0"></a>
# 2.0.0 (YYYY-MM-DD)


### Features

* init repo SHA
* **package-3:** Add baz feature SHA


### BREAKING CHANGES

* **package-3:** yup
,
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="2.0.0"></a>
# 2.0.0 (YYYY-MM-DD)


### Features

* init repo SHA
,
  # Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="2.0.0"></a>
# 2.0.0 (YYYY-MM-DD)


### Features

* init repo SHA
,
]
`;

exports[`lerna publish replaces file: specifier with local version before npm publish but after git commit: committed 1`] = `
v2.0.0

diff --git a/lerna.json b/lerna.json
index SHA..SHA 100644
--- a/lerna.json
+++ b/lerna.json
@@ -2 +2 @@
-  "version": "1.0.0"
+  "version": "2.0.0"
diff --git a/packages/package-1/package.json b/packages/package-1/package.json
index SHA..SHA 100644
--- a/packages/package-1/package.json
+++ b/packages/package-1/package.json
@@ -3 +3 @@
-	"version": "1.0.0",
+	"version": "2.0.0",
diff --git a/packages/package-2/package.json b/packages/package-2/package.json
index SHA..SHA 100644
--- a/packages/package-2/package.json
+++ b/packages/package-2/package.json
@@ -3 +3 @@
-  "version": "1.0.0",
+  "version": "2.0.0",
diff --git a/packages/package-3/package.json b/packages/package-3/package.json
index SHA..SHA 100644
--- a/packages/package-3/package.json
+++ b/packages/package-3/package.json
@@ -3 +3 @@
-  "version": "1.0.0",
+  "version": "2.0.0",
diff --git a/packages/package-4/package.json b/packages/package-4/package.json
index SHA..SHA 100644
--- a/packages/package-4/package.json
+++ b/packages/package-4/package.json
@@ -3 +3 @@
-  "version": "1.0.0",
+  "version": "2.0.0",
diff --git a/packages/package-5/package.json b/packages/package-5/package.json
index SHA..SHA 100644
--- a/packages/package-5/package.json
+++ b/packages/package-5/package.json
@@ -3 +3 @@
-  "version": "1.0.0",
+  "version": "2.0.0",
diff --git a/packages/package-7/package.json b/packages/package-7/package.json
index SHA..SHA 100644
--- a/packages/package-7/package.json
+++ b/packages/package-7/package.json
@@ -3 +3 @@
-  "version": "1.0.0",
+  "version": "2.0.0",
`;

exports[`lerna publish replaces file: specifier with local version before npm publish but after git commit: unstaged 1`] = `
diff --git a/packages/package-1/package.json b/packages/package-1/package.json
index SHA..SHA 100644
--- a/packages/package-1/package.json
+++ b/packages/package-1/package.json
@@ -7 +7,2 @@
-	}
+	},
+	"gitHead": "GIT_HEAD"
diff --git a/packages/package-2/package.json b/packages/package-2/package.json
index SHA..SHA 100644
--- a/packages/package-2/package.json
+++ b/packages/package-2/package.json
@@ -5,2 +5,3 @@
-    "package-1": "file:../package-1"
-  }
+    "package-1": "^2.0.0"
+  },
+  "gitHead": "GIT_HEAD"
diff --git a/packages/package-3/package.json b/packages/package-3/package.json
index SHA..SHA 100644
--- a/packages/package-3/package.json
+++ b/packages/package-3/package.json
@@ -5,2 +5,3 @@
-    "package-2": "file:../package-2"
-  }
+    "package-2": "^2.0.0"
+  },
+  "gitHead": "GIT_HEAD"
diff --git a/packages/package-4/package.json b/packages/package-4/package.json
index SHA..SHA 100644
--- a/packages/package-4/package.json
+++ b/packages/package-4/package.json
@@ -5,2 +5,3 @@
-    "package-3": "file:../package-3"
-  }
+    "package-3": "^2.0.0"
+  },
+  "gitHead": "GIT_HEAD"
diff --git a/packages/package-5/package.json b/packages/package-5/package.json
index SHA..SHA 100644
--- a/packages/package-5/package.json
+++ b/packages/package-5/package.json
@@ -5,3 +5,4 @@
-    "package-4": "file:../package-4",
-    "package-6": "file:../package-6"
-  }
+    "package-4": "^2.0.0",
+    "package-6": "^1.0.0"
+  },
+  "gitHead": "GIT_HEAD"
`;

exports[`lerna publish silences lifecycle scripts with --loglevel=silent 1`] = `

Changes:
 - package-1: 1.0.0 => 1.1.0
 - package-2: 1.0.0 => 1.1.0

preversion-root
preversion-package-1
version-package-1
version-root
postversion-package-1
postversion-root
`;

exports[`lerna publish updates all transitive dependents 1`] = `
Array [
  Object {
    change: true,
    description: no local dependencies, four local dependents (three transitive),
    gitHead: GIT_HEAD,
    name: package-1,
    version: 2.0.0,
  },
  Object {
    dependencies: Object {
      package-1: ^2.0.0,
    },
    description: one local dependency, one direct dependent, no transitive dependencies,
    gitHead: GIT_HEAD,
    name: package-2,
    version: 2.0.0,
  },
  Object {
    dependencies: Object {
      package-2: ^2.0.0,
    },
    description: one local dependency, one direct dependent, one transitive dependency,
    gitHead: GIT_HEAD,
    name: package-3,
    version: 2.0.0,
  },
  Object {
    dependencies: Object {
      package-3: ^2.0.0,
    },
    description: one local dependency, one direct dependent, two transitive dependencies,
    gitHead: GIT_HEAD,
    name: package-4,
    version: 2.0.0,
  },
  Object {
    dependencies: Object {
      package-4: ^2.0.0,
    },
    description: one local dependency, no dependents, three transitive dependencies,
    gitHead: GIT_HEAD,
    name: package-5,
    version: 2.0.0,
  },
]
`;

exports[`lerna publish updates fixed versions: commit 1`] = `
v1.0.1

`;

exports[`lerna publish updates fixed versions: packages 1`] = `
Array [
  Object {
    gitHead: GIT_HEAD,
    name: package-1,
    version: 1.0.1,
  },
  Object {
    dependencies: Object {
      package-1: ^1.0.1,
    },
    gitHead: GIT_HEAD,
    name: package-2,
    version: 1.0.1,
  },
  Object {
    devDependencies: Object {
      package-2: ^1.0.1,
    },
    gitHead: GIT_HEAD,
    name: package-3,
    peerDependencies: Object {
      package-2: ^1.0.0,
    },
    version: 1.0.1,
  },
  Object {
    dependencies: Object {
      package-1: ^0.0.0,
    },
    gitHead: GIT_HEAD,
    name: package-4,
    version: 1.0.1,
  },
  Object {
    dependencies: Object {
      package-1: ^1.0.1,
    },
    name: package-5,
    private: true,
    version: 1.0.1,
  },
]
`;

exports[`lerna publish updates fixed versions: stdout 1`] = `

Changes:
 - package-1: 1.0.0 => 1.0.1
 - package-2: 1.0.0 => 1.0.1
 - package-3: 1.0.0 => 1.0.1
 - package-4: 1.0.0 => 1.0.1
 - package-5: 1.0.0 => 1.0.1 (private)

`;

exports[`lerna publish updates independent versions: commit 1`] = `
Publish

 - package-1@2.0.0
 - package-2@3.0.0
 - package-3@4.0.0
 - package-4@5.0.0
 - package-5@6.0.0

`;

exports[`lerna publish updates independent versions: packages 1`] = `
Array [
  Object {
    gitHead: GIT_HEAD,
    name: package-1,
    version: 2.0.0,
  },
  Object {
    dependencies: Object {
      package-1: ^2.0.0,
    },
    gitHead: GIT_HEAD,
    name: package-2,
    version: 3.0.0,
  },
  Object {
    devDependencies: Object {
      package-2: ^3.0.0,
    },
    gitHead: GIT_HEAD,
    name: package-3,
    version: 4.0.0,
  },
  Object {
    dependencies: Object {
      package-1: ^0.0.0,
    },
    gitHead: GIT_HEAD,
    name: package-4,
    version: 5.0.0,
  },
  Object {
    dependencies: Object {
      package-3: ^4.0.0,
    },
    name: package-5,
    private: true,
    version: 6.0.0,
  },
]
`;

exports[`lerna publish updates independent versions: stdout 1`] = `

Changes:
 - package-1: 1.0.0 => 2.0.0
 - package-2: 2.0.0 => 3.0.0
 - package-3: 3.0.0 => 4.0.0
 - package-4: 4.0.0 => 5.0.0
 - package-5: 5.0.0 => 6.0.0 (private)

`;

exports[`lerna publish uses default suffix with canary flag: stdout 1`] = `

Changes:
 - package-1: 1.0.0 => 1.1.0-alpha.SHA
 - package-2: 1.0.0 => 1.1.0-alpha.SHA
 - package-3: 1.0.0 => 1.1.0-alpha.SHA
 - package-4: 1.0.0 => 1.1.0-alpha.SHA
 - package-5: 1.0.0 => 1.1.0-alpha.SHA (private)

`;
